Dapr Traffic Control

amolenk opened the pull request:
## Summary

Adds the traffic control sample application to the Dapr book and consolidates all eShop info in a single chapter.

### Internal preview(s)

https://review.docs.microsoft.com/en-us/dotnet/architecture/dapr-for-net-developers/?branch=pr-en-us-23490

robvet committed:
We lost history right before we published. Updating msdate metadata values to get added back as a contributor on each chapter

robvet committed:
Removed tracing exporters from Ch 2

robvet committed:
Merge remote-tracking branch 'upstream/master' into master

robvet committed:
Merge branch 'ch3-updates' into master

amolenk committed:
Update TOC

amolenk committed:
Add sample chapter

amolenk committed:
Consolidated eShop content into separate chapter

amolenk committed:
Reference application summary

amolenk committed:
Benefits in own section

amolenk committed:
Getting started todos

amolenk committed:
Formatting

amolenk committed:
State sample

EdwinVW committed:
Add sample-application chapter

EdwinVW committed:
Merge remote-tracking branch 'origin/dapr/trafficcontrol' into dapr/trafficcontrol

EdwinVW committed:
Fix mdlint warning

EdwinVW committed:
Minor rephrase

EdwinVW committed:
Update  pub/sub chapter

EdwinVW committed:
Fix mdlint warnings

amolenk committed:
Naming and summary

amolenk committed:
Update images

amolenk committed:
Remove SDK note

amolenk committed:
Change title for consistency

amolenk committed:
Update image syntax

amolenk committed:
Rename title for consistency

EdwinVW committed:
Add service invocation

EdwinVW committed:
Fix typo + rephrase benefits

EdwinVW committed:
Merge branch 'dapr/trafficcontrol' of https://github.com/amolenk/docs into dapr/trafficcontrol

EdwinVW committed:
Fix typo

EdwinVW committed:
Change image tags

EdwinVW committed:
Fix mdlint warnings

EdwinVW committed:
Add note about instructions to run the sample app

EdwinVW committed:
Add bindings chapter

robvet committed:
review of state management

robvet committed:
updates to pub/sub. added graphic to state management

IEvangelist commented on docs/architecture/dapr-for-net-developers/reference-application.md:
```diff
@@ -52,38 +52,606 @@ As focus of the eShopOnDapr reference application is on Dapr, the original appli
+{
+    var string message = CreateEmailBody(notification);
+    var metadata = new Dictionary<string, string>
+    {
+        {"emailFrom", "eShopOn@dapr.io"},
+        {"emailTo", notification.UserName},
+        {"subject", $"Your eShopOnDapr order #{notification.Order.Id}"}
+    };
```
```suggestion
    {
        ["emailFrom"] = "eShopOn@dapr.io",
        ["emailTo" = notification.UserName,
        ["subject"] = $"Your eShopOnDapr order #{notification.Order.Id}"
    };
```

For consistency with the other dictionary collection initializer usages in these articles.

EdwinVW committed:
Add benefits

EdwinVW committed:
Fix link + mdlint warning

EdwinVW committed:
Update secrets management

EdwinVW committed:
Update docs/architecture/dapr-for-net-developers/bindings.md

Co-authored-by: David Pine <david.pine@microsoft.com>

EdwinVW committed:
Update docs/architecture/dapr-for-net-developers/bindings.md

Co-authored-by: David Pine <david.pine@microsoft.com>

EdwinVW committed:
Update building block name

robvet committed:
Merge branch 'dapr/trafficcontrol' into dapr/P0/review

robvet committed:
Review of service invocation

robvet committed:
reviewed bindings

EdwinVW committed:
Add observability

EdwinVW committed:
Update  images and codesnippets to traffic control

EdwinVW committed:
Add hosting paragraph + fix typo

EdwinVW committed:
Fix log entries

amolenk committed:
Add actors text and renumber

amolenk committed:
Add actor diagrams and fix markdown

IEvangelist commented:
Hi @amolenk and @EdwinVW - any updates on this PR?

amolenk commented:
Hi @IEvangelist - I'll be adding content on actors over the weekend. @robvet will also be doing more internal review this weekend. After that it'll be ready for final review/approval.

amolenk committed:
Actors

robvet committed:
wip: secrets management review

robvet committed:
merge conflicts with main

robvet committed:
Merge branch 'dapr/trafficcontrol' into dapr/P0/review

amolenk committed:
Fix links

robvet committed:
secrets management review

robvet committed:
review completed for observability. Files datetime updated

robvet committed:
completed review of eshopondapr reference app

robvet committed:
sample app chapter reviewed

amolenk committed:
Fix typo

EdwinVW committed:
Update images

EdwinVW committed:
Update eShopOnDapr solution image

amolenk committed:
Add extra note for HTTPS

amolenk committed:
Merge branch 'dapr/trafficcontrol' of https://github.com/amolenk/docs into dapr/trafficcontrol

amolenk committed:
Merge branch 'main' into dapr/trafficcontrol

EdwinVW committed:
Update images

EdwinVW committed:
Merge branch 'dapr/trafficcontrol' of https://github.com/amolenk/docs into dapr/trafficcontrol

EdwinVW committed:
Fix image captions

EdwinVW committed:
Update invoke-actor-method image

amolenk committed:
Actors in reference application

amolenk committed:
Merge branch 'dapr/trafficcontrol' of https://github.com/amolenk/docs into dapr/trafficcontrol

amolenk committed:
Merge branch 'main' into dapr/trafficcontrol

amolenk committed:
Remove obsolete file

robvet committed:
Fixed lint error

amolenk committed:
Merge branch 'main' into dapr/trafficcontrol

amolenk committed:
Update edition number

amolenk committed:
Add name resolution components

robvet committed:
final review on secrets management

robvet committed:
completed review for observability chapter

robvet committed:
completed review of reference application chapter

robvet committed:
completed review of sample application

robvet committed:
Corrected note captions

amolenk committed:
Merge branch 'dapr/trafficcontrol' into dapr/P0/review

amolenk committed:
Merge pull request #300 from amolenk/dapr/P0/review

Building block, sample, and reference app review

amolenk committed:
Bindings review

amolenk committed:
Observability review

amolenk committed:
Pub/sub review

amolenk committed:
Secrets review

amolenk committed:
Service invocation review

EdwinVW committed:
Update sample application images

EdwinVW committed:
Add image caption

amolenk committed:
Merge

amolenk committed:
State management review

amolenk committed:
Fix alt text

amolenk commented:
@IEvangelist Hi David, this PR is now ready for review.

Unfortunately, the MSDocs build verifier reports an error at the moment: `No redirection is found for 'docs/architecture/dapr-for-net-developers/secrets.md`. The `secrets.md` file that the error message refers to used to exist in the repo, but has now been renamed to `secrets-management.md`. I have no idea why this breaks the build, can you take a look at it?

IEvangelist commented:
> @IEvangelist David Pine FTE Hi David, this PR is now ready for review.
> 
> Unfortunately, the MSDocs build verifier reports an error at the moment: `No redirection is found for 'docs/architecture/dapr-for-net-developers/secrets.md`. The `secrets.md` file that the error message refers to used to exist in the repo, but has now been renamed to `secrets-management.md`. I have no idea why this breaks the build, can you take a look at it?

Hi @amolenk - thanks for the update on this. Yes, so the issue is that since the file was published at one point in time, and now renamed - the URL for the page is different. As such anyone who was using the URL, internal or external (bookmarking or whatever) will be broken and land on a 404 page. To fix this, as file names change - we need to also maintain the mapping for redirects. Does that make sense?

All you need to do is add an entry to the end _.openpublishing.redirections.json_ file at the root of the repo - something similar to the following:

```json
{
  "source_path": "docs/architecture/dapr-for-net-developers/secrets.md",
  "redirect_url": "/dotnet/architecture/dapr-for-net-developers/secrets-management"
}
```

Where the source path is the original MD file, and the redirect URL is the route to the new file.

amolenk committed:
Add redirect for deleted secrets.md

amolenk commented:
Thanks @IEvangelist! All checks have passed now

IEvangelist commented:
> Thanks @IEvangelist David Pine FTE! All checks have passed now

Hi @amolenk - I'm going to review this first thing Monday morning. Been buried, thank you for your patience.

IEvangelist commented on docs/architecture/dapr-for-net-developers/secrets-management.md:
```diff
@@ -463,85 +463,270 @@ You specify scopes per secret store. In the above example, the secret store is n
+                throw new InvalidOperationException("Invalid license-key specified.");
+            }
+
+            int fine = 9; // default administration fee
+            if (violationInKmh < 5 )
+            {
+                fine += 18;
+            }
+            else if (violationInKmh >= 5 && violationInKmh < 10 )
+            {
+                fine += 31;
+            }
+
+            // ...
+            
+            else if (violationInKmh == 35)
+            {
+                fine += 372;
+            }
+            else
+            {
+                // violation above 35 KMh will be determined by the prosecutor
+                return 0;
+            }
+
+            return fine;
```
```suggestion
            int fine = 9; // default administration fee    
            return violationInKmh switch
            {
                < 5 => fine + 18,
                >= 5 and < 10 => fine + 31,

                // ...

                35 => fine + 372,
                // violation above 35 KMh will be determined by the prosecutor
                > 35 => 0,
                _ => fine
            };
```

IEvangelist commented on docs/architecture/dapr-for-net-developers/service-invocation.md:
```diff
@@ -228,217 +228,114 @@ catch (InvocationException ex)
+## Name resolution components
 
-The original [eShopOnContainers](https://github.com/dotnet-architecture/eShopOnContainers) microservice reference architecture from Microsoft used a mix of HTTP/REST and gRPC services. The use of gRPC was limited to communication between an [aggregator service](../cloud-native/service-to-service-communication.md#service-aggregator-pattern) and core back-end services. Figure 6-2 show the architect ...
+At the time of this writing, Dapr provides support for the following name resolution components:
```
```suggestion
At the time of writing, Dapr provides support for the following name resolution components:
```

IEvangelist commented on docs/architecture/dapr-for-net-developers/service-invocation.md:
````diff
@@ -228,217 +228,114 @@ catch (InvocationException ex)
+ ```csharp
+ [Topic("pubsub", "speedingviolations")]
+ [Route("collectfine")]
+ [HttpPost()]
````
```suggestion
 [HttpPost]
```

IEvangelist commented on docs/architecture/dapr-for-net-developers/state-management.md:
```diff
@@ -161,7 +160,7 @@ For bulk operations, Dapr will submit each key/value pair update as a separate r
 ## Use the Dapr .NET SDK
 
-The Dapr .NET SDK provides language-specific support for .NET Core platform. Developers can use the `DaprClient` class introduced in [chapter 3](getting-started.md) to read and write data. The following example shows how to use the `DaprClient.GetStateAsync<TValue>` method to read data from a state store. The method expects the store name, `statestore`, and key, `AMS`, as parameters:
+The Dapr .NET SDK provides language-specific support for the .NET Core platform. Developers can use the `DaprClient` class introduced in [chapter 3](getting-started.md) to read and write data. The following example shows how to use the `DaprClient.GetStateAsync<TValue>` method to read data from a state store. The method expects the store name, `statestore`, and key, `AMS`, as parameters:
```
```suggestion
The Dapr .NET SDK provides language-specific support for the .NET platform. Developers can use the `DaprClient` class introduced in [chapter 3](getting-started.md) to read and write data. The following example shows how to use the `DaprClient.GetStateAsync<TValue>` method to read data from a state store. The method expects the store name, `statestore`, and key, `AMS`, as parameters:
```

IEvangelist reviewed (changes requested):
Hi again @amolenk - Overall, I love these changes / updates. This seems to have been a huge effort - thank you. With that said, my review has yielded a bit more work. Please let me know when you've addressed the feedback. Thanks again.

EdwinVW commented:
Thanks for the review @IEvangelist. Glad you like the changes.
We will pick this up a.s.a.p. and keep you in the loop about the progress.

EdwinVW committed:
Apply suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>

IEvangelist commented:
> Thanks for the review @IEvangelist David Pine FTE. Glad you like the changes.
> We will pick this up a.s.a.p. and keep you in the loop about the progress.

Sounds good, I am curious about the link to the traffic control sample app. I would have assumed that this would have been part of the `dotnet-architecture` repo? @nishanil is this something we should do?

EdwinVW committed:
Apply suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>

EdwinVW committed:
Merge branch 'main' into dapr/trafficcontrol

amolenk committed:
Apply Actor suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>

amolenk committed:
Apply Reference Application suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>

amolenk committed:
Merge branch 'main' into dapr/trafficcontrol

amolenk committed:
Merge branch 'dapr/trafficcontrol' of https://github.com/amolenk/docs into dapr/trafficcontrol

amolenk committed:
Actor review

amolenk commented on docs/architecture/dapr-for-net-developers/actors.md:
```diff
@@ -0,0 +1,630 @@
+
+## Summary
+
+The Dapr actors building block makes it easier to write correct concurrent systems. Actors are small units of state and logic. They use a turn-based access model which saves you from having to write error-prone thread-safe code. Actors are created implicitly and are silently unloaded from memory when no operations are performed. Any state stored in the actor is automatically persisted and loaded  ...
```
Rewritten the sentence to "They use a turn-based access model which saves you from having to use locking mechanisms to write thread-safe code."

EdwinVW committed:
Update docs/architecture/dapr-for-net-developers/bindings.md

Co-authored-by: David Pine <david.pine@microsoft.com>

amolenk committed:
Reference app review

amolenk committed:
Merge branch 'dapr/trafficcontrol' of https://github.com/amolenk/docs into dapr/trafficcontrol

amolenk commented on docs/architecture/dapr-for-net-developers/reference-application.md:
```diff
@@ -2,92 +2,988 @@
+1. The system processes the payment.
+1. The system ships the order.
+
+The process is implemented using a single `OrderingProcessActor` actor type. Here's the interface for the actor:
```
Fixed

amolenk commented on docs/architecture/dapr-for-net-developers/reference-application.md:
```diff
@@ -2,92 +2,988 @@
+{
+    _logger.LogInformation("Received {Actor}[{ActorId}] reminder: {Reminder}", nameof(OrderingProcessActor), OrderId, reminderName);
+
+    switch (reminderName)
```
Yes, changed implementation to use `switch` expression

amolenk commented:
Hi @IEvangelist, I believe we've addressed all your review comments. Great suggestions BTW, thanks!

@msfussell and @robvet would like to have a final look at the new actors chapter. After that, it'll be ready to merge.

IEvangelist commented on docs/architecture/dapr-for-net-developers/actors.md:
```diff
@@ -0,0 +1,636 @@
+public async Task<int> GetScoreAsync()
+{
+    var scoreValue = await StateManager.TryGetStateAsync<int>("score");
+    if (scoreValue.HasValue)
+    {
+        return scoreValue.Value;
+    }
+
+    return 0;
```
If this is a `Nullable<int>` (`int?`), maybe simplify the return to this?

```suggestion
    return scoreValue.GetValueOrDefault();
```

amolenk replied:
`GetStateAsync` returns an instance of `Dapr.Actors.Runtime.ConditionalValue<TValue>` which only has `HasValue()` and `Value` members

IEvangelist commented on docs/architecture/dapr-for-net-developers/service-invocation.md:
```diff
@@ -105,7 +105,7 @@ Alternatively, you can configure a custom endpoint in the call to `DaprClient.Cr
 var httpClient = DaprClient.CreateHttpClient(daprEndpoint = "localhost:4000");
```
```suggestion
var httpClient = DaprClient.CreateInvokeHttpClient(daprEndpoint = "localhost:4000");
```

See, https://github.com/dotnet/docs/pull/24652 - I'm curious if this comes from an extension method or what? But we should figure that out and address this.

amolenk replied:
Not an extension method, just a typo unfortunately :-(
Search through the entire text, #24652 fixed them all.

IEvangelist commented on docs/architecture/dapr-for-net-developers/service-invocation.md:
````diff
@@ -105,7 +105,7 @@ Alternatively, you can configure a custom endpoint in the call to `DaprClient.Cr
+You can also directly set the base address by specifying the application ID. Doing so enables relative URIs when making a call:
 
 ```csharp
 var httpClient = DaprClient.CreateHttpClient("orderservice");
````
```suggestion
var httpClient = DaprClient.CreateInvokeHttpClient("orderservice");
```

IEvangelist reviewed (commented):
Hi @amolenk and @EdwinVW - We need to look into a recent issue that this PR seems to also include, see https://github.com/dotnet/docs/pull/24652

amolenk committed:
Merge branch 'main' into dapr/trafficcontrol

EdwinVW committed:
Update cover

msfussell commented on docs/architecture/dapr-for-net-developers/actors.md:
```diff
@@ -0,0 +1,636 @@
+
+Unfortunately, using explicit locking mechanisms is error-prone. They can easily lead to deadlocks and can have serious impact on performance.
+
+Thanks to the turn-based access model, you don't need to worry about multiple threads with actors, making it much easier to concurrent systems. The following actor example closely mirrors the code from the previous sample, but doesn't require any locking mechanisms to be correct:
```
typo "...making it much easier to concurrent systems."

msfussell commented on docs/architecture/dapr-for-net-developers/actors.md:
```diff
@@ -0,0 +1,636 @@
+- MongoDB
+- PostgreSQL
+- Redis
+- SQL Server
```
There are 7 in total as of today. See https://docs.dapr.io/reference/components-reference/supported-state-stores/

msfussell commented on docs/architecture/dapr-for-net-developers/actors.md:
```diff
@@ -0,0 +1,636 @@
+
+## Use the Dapr .NET SDK
+
+You can create an actor model implementation using only HTTP/gRPC calls. However, it's much more convenient to use the language specific Dapr SDKs. At the time of writing, both the .NET and Java SDKs provide extensive support for working with actors.
```
Java, .NET and Python support Actors.  Need to give Python some love....

msfussell commented on docs/architecture/dapr-for-net-developers/actors.md:
```diff
@@ -0,0 +1,636 @@
+The actors endpoints are necessary because the Dapr sidecar calls the application to host and interact with actor instances.
+
+> [!IMPORTANT]
+> By default, a generated `Startup` class contains an `app.UseHttpsRedirection` call to redirect client to the HTTPS endpoint. This will not work with actors and must be removed. By design, a Dapr sidecar sends requests over unencrypted HTTP by default. The HTTPS middleware will block these requests when enabled.
```
The code generated by VS2019 for ASP.NET web apps I presume
